File: HisabCore/Sources/HisabCore/GenericBank/ChainInterpreter.swift
import Foundation

/// Which cell means what in a bank-statement body row. Exactly one amount
/// convention applies: separate debit/credit columns, a single signed amount,
/// an unsigned amount qualified by a DR/CR column, or an unsigned amount whose
/// direction only the balance chain can reveal (`amountIsUnsigned`).
public struct ColumnMapping: Equatable, Sendable {
    public var date: Int
    public var narration: Int
    public var reference: Int?
    public var debit: Int?
    public var credit: Int?
    public var amount: Int?
    public var drcr: Int?
    public var balance: Int
    public var amountIsUnsigned: Bool
    public var dateFormat: String

    public init(date: Int, narration: Int, reference: Int? = nil,
                debit: Int? = nil, credit: Int? = nil,
                amount: Int? = nil, drcr: Int? = nil,
                balance: Int, amountIsUnsigned: Bool = false, dateFormat: String) {
        self.date = date
        self.narration = narration
        self.reference = reference
        self.debit = debit
        self.credit = credit
        self.amount = amount
        self.drcr = drcr
        self.balance = balance
        self.amountIsUnsigned = amountIsUnsigned
        self.dateFormat = dateFormat
    }
}

public enum ChainOutcome: Sendable, Equatable {
    case validated([ParsedTransaction])
    case broken(rowIndex: Int, detail: String)
}

/// Applies a mapping to body rows and accepts the result only when the running
/// balance closes to the paisa on every row. When the columns assert a
/// direction, the chain must agree — disagreement is a break, not a warning.
public enum ChainInterpreter {
    public static func interpret(rows: [[String]], mapping: ColumnMapping,
                                 openingBalancePaise: Int64? = nil) -> ChainOutcome {
        let first = run(rows: rows, mapping: mapping,
                        openingBalance: openingBalancePaise, openingDirection: .debit)
        if mapping.amountIsUnsigned, mapping.drcr == nil, openingBalancePaise == nil {
            // Row 1 has no predecessor to anchor its direction; try both and let
            // the rest of the chain arbitrate.
            let second = run(rows: rows, mapping: mapping,
                             openingBalance: nil, openingDirection: .credit)
            switch (first, second) {
            case (.validated(let a), .validated(let b)):
                return a == b ? .validated(a)
                              : .broken(rowIndex: 0, detail: "ambiguous opening direction")
            case (.validated, .broken): return first
            case (.broken, .validated): return second
            case (.broken, .broken): return first
            }
        }
        return first
    }

    private static func run(rows: [[String]], mapping: ColumnMapping,
                            openingBalance: Int64?, openingDirection: Direction) -> ChainOutcome {
        let formatter = DateFormatter()
        formatter.calendar = YearMonth.istCalendar
        formatter.timeZone = YearMonth.istCalendar.timeZone
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = mapping.dateFormat

        var transactions: [ParsedTransaction] = []
        var previousBalance = openingBalance

        for (index, row) in rows.enumerated() {
            func cell(_ column: Int?) -> String {
                guard let column, column >= 0, column < row.count else { return "" }
                return row[column]
            }

            let dateText = cell(mapping.date)
            guard let date = formatter.date(from: dateText) else {
                return .broken(rowIndex: index, detail: "unreadable date '\(dateText)'")
            }
            let balanceText = cell(mapping.balance)
            guard let balance = Money.signedPaise(fromDecimalString: balanceText) else {
                return .broken(rowIndex: index, detail: "unreadable balance '\(balanceText)'")
            }

            var amount: Int64
            var assertedDirection: Direction?
            if let debitColumn = mapping.debit, let creditColumn = mapping.credit {
                let debitText = cell(debitColumn)
                let creditText = cell(creditColumn)
                switch (debitText.isEmpty, creditText.isEmpty) {
                case (false, true):
                    guard let value = Money.signedPaise(fromDecimalString: debitText) else {
                        return .broken(rowIndex: index, detail: "unreadable debit '\(debitText)'")
                    }
                    amount = value
                    assertedDirection = .debit
                case (true, false):
                    guard let value = Money.signedPaise(fromDecimalString: creditText) else {
                        return .broken(rowIndex: index, detail: "unreadable credit '\(creditText)'")
                    }
                    amount = value
                    assertedDirection = .credit
                default:
                    return .broken(rowIndex: index,
                                   detail: "expected exactly one of debit/credit, got '\(debitText)'/'\(creditText)'")
                }
            } else if let amountColumn = mapping.amount {
                let amountText = cell(amountColumn)
                guard let value = Money.signedPaise(fromDecimalString: amountText) else {
                    return .broken(rowIndex: index, detail: "unreadable amount '\(amountText)'")
                }
                if let drcrColumn = mapping.drcr {
                    switch cell(drcrColumn).lowercased() {
                    case "dr": assertedDirection = .debit
                    case "cr": assertedDirection = .credit
                    default:
                        return .broken(rowIndex: index, detail: "unreadable DR/CR '\(cell(drcrColumn))'")
                    }
                    amount = abs(value)
                } else if mapping.amountIsUnsigned {
                    amount = abs(value)
                    assertedDirection = nil
                } else {
                    amount = abs(value)
                    assertedDirection = value < 0 ? .debit : .credit
                }
            } else {
                return .broken(rowIndex: index, detail: "mapping declares no amount columns")
            }

            guard amount > 0 else {
                return .broken(rowIndex: index, detail: "non-positive amount")
            }

            let direction: Direction
            if let previous = previousBalance {
                if previous - amount == balance {
                    direction = .debit
                } else if previous + amount == balance {
                    direction = .credit
                } else {
                    return .broken(rowIndex: index,
                                   detail: "balance chain break: \(previous) ±\(amount) ≠ \(balance)")
                }
                if let asserted = assertedDirection, asserted != direction {
                    return .broken(rowIndex: index,
                                   detail: "columns say \(asserted.rawValue), chain says \(direction.rawValue)")
                }
            } else {
                direction = assertedDirection ?? openingDirection
            }

            let narration = cell(mapping.narration)
            let referenceCell = cell(mapping.reference)
            let reference = referenceCell.isEmpty
                ? SyntheticRef.make(balancePaise: balance, date: date, amountPaise: amount)
                : referenceCell
            transactions.append(ParsedTransaction(date: date, amountPaise: amount,
                                                  direction: direction,
                                                  counterparty: narration,
                                                  reference: reference,
                                                  narration: narration))
            previousBalance = balance
        }

        return .validated(transactions)
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/generic_bank/chain_interpreter.dart
/// Balance-chain-validated column interpretation. Port of
/// ChainInterpreter.swift — same semantics, including the two-pass opening
/// direction and the columns-vs-chain agreement rule.
library;

import '../domain.dart';
import '../money.dart';
import 'statement_date.dart';
import 'synthetic_ref.dart';

class ColumnMapping {
  final int date;
  final int narration;
  final int? reference;
  final int? debit;
  final int? credit;
  final int? amount;
  final int? drcr;
  final int balance;
  final bool amountIsUnsigned;
  final String dateFormat;
  /// Regexes with one capture group, tried against the narration when the
  /// reference cell is empty; first non-empty capture wins, else synthetic.
  final List<String> referencePatterns;

  const ColumnMapping({
    required this.date,
    required this.narration,
    this.reference,
    this.debit,
    this.credit,
    this.amount,
    this.drcr,
    required this.balance,
    this.amountIsUnsigned = false,
    required this.dateFormat,
    this.referencePatterns = const [],
  });
}

sealed class ChainOutcome {
  const ChainOutcome();
}

class ChainValidated extends ChainOutcome {
  final List<ParsedTransaction> transactions;
  const ChainValidated(this.transactions);
}

class ChainBroken extends ChainOutcome {
  final int rowIndex;
  final String detail;
  const ChainBroken(this.rowIndex, this.detail);
}

class ChainInterpreter {
  static ChainOutcome interpret({
    required List<List<String>> rows,
    required ColumnMapping mapping,
    int? openingBalancePaise,
  }) {
    final first = _run(rows, mapping, openingBalancePaise, Direction.debit);
    if (mapping.amountIsUnsigned &&
        mapping.drcr == null &&
        openingBalancePaise == null) {
      // Row 1 has no predecessor to anchor its direction; try both and let
      // the rest of the chain arbitrate.
      final second = _run(rows, mapping, null, Direction.credit);
      if (first is ChainValidated && second is ChainValidated) {
        return _sameTxns(first.transactions, second.transactions)
            ? first
            : const ChainBroken(0, 'ambiguous opening direction');
      }
      if (first is ChainValidated) return first;
      if (second is ChainValidated) return second;
      return first;
    }
    return first;
  }

  static bool _sameTxns(List<ParsedTransaction> a, List<ParsedTransaction> b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  static ChainOutcome _run(List<List<String>> rows, ColumnMapping mapping,
      int? openingBalance, Direction openingDirection) {
    final referenceRegexes = <RegExp>[];
    for (final pattern in mapping.referencePatterns) {
      try {
        referenceRegexes.add(RegExp(pattern));
      } on FormatException {
        // ignore malformed pattern — spec lint should catch it
      }
    }

    final transactions = <ParsedTransaction>[];
    var previousBalance = openingBalance;

    for (var index = 0; index < rows.length; index++) {
      final row = rows[index];
      String cell(int? column) =>
          (column == null || column < 0 || column >= row.length)
              ? ''
              : row[column];

      final dateText = cell(mapping.date);
      final date = StatementDate.tryParse(dateText, mapping.dateFormat);
      if (date == null) {
        return ChainBroken(index, "unreadable date '$dateText'");
      }
      final balanceText = cell(mapping.balance);
      final balance = Money.signedPaise(balanceText);
      if (balance == null) {
        return ChainBroken(index, "unreadable balance '$balanceText'");
      }

      int amount;
      Direction? assertedDirection;
      final debitColumn = mapping.debit;
      final creditColumn = mapping.credit;
      final amountColumn = mapping.amount;
      if (debitColumn != null && creditColumn != null) {
        final debitText = cell(debitColumn);
        final creditText = cell(creditColumn);
        if (debitText.isNotEmpty && creditText.isEmpty) {
          final value = Money.signedPaise(debitText);
          if (value == null) {
            return ChainBroken(index, "unreadable debit '$debitText'");
          }
          amount = value;
          assertedDirection = Direction.debit;
        } else if (debitText.isEmpty && creditText.isNotEmpty) {
          final value = Money.signedPaise(creditText);
          if (value == null) {
            return ChainBroken(index, "unreadable credit '$creditText'");
          }
          amount = value;
          assertedDirection = Direction.credit;
        } else {
          return ChainBroken(index,
              "expected exactly one of debit/credit, got '$debitText'/'$creditText'");
        }
      } else if (amountColumn != null) {
        final amountText = cell(amountColumn);
        final value = Money.signedPaise(amountText);
        if (value == null) {
          return ChainBroken(index, "unreadable amount '$amountText'");
        }
        final drcrColumn = mapping.drcr;
        if (drcrColumn != null) {
          switch (cell(drcrColumn).toLowerCase()) {
            case 'dr':
              assertedDirection = Direction.debit;
            case 'cr':
              assertedDirection = Direction.credit;
            default:
              return ChainBroken(
                  index, "unreadable DR/CR '${cell(drcrColumn)}'");
          }
          amount = value.abs();
        } else if (mapping.amountIsUnsigned) {
          amount = value.abs();
          assertedDirection = null;
        } else {
          amount = value.abs();
          assertedDirection =
              value < 0 ? Direction.debit : Direction.credit;
        }
      } else {
        return ChainBroken(index, 'mapping declares no amount columns');
      }

      if (amount <= 0) {
        return ChainBroken(index, 'non-positive amount');
      }

      Direction direction;
      final previous = previousBalance;
      if (previous != null) {
        if (previous - amount == balance) {
          direction = Direction.debit;
        } else if (previous + amount == balance) {
          direction = Direction.credit;
        } else {
          return ChainBroken(
              index, 'balance chain break: $previous ±$amount ≠ $balance');
        }
        if (assertedDirection != null && assertedDirection != direction) {
          return ChainBroken(index,
              'columns say ${assertedDirection.name}, chain says ${direction.name}');
        }
      } else {
        direction = assertedDirection ?? openingDirection;
      }

      final narration = cell(mapping.narration);
      final referenceCell = cell(mapping.reference);
      String reference;
      if (referenceCell.isNotEmpty) {
        reference = referenceCell;
      } else {
        final extracted = _extractReference(narration, referenceRegexes);
        reference = extracted ??
            SyntheticRef.make(
                balancePaise: balance, date: date, amountPaise: amount);
      }
      transactions.add(ParsedTransaction(
        date: date,
        amountPaise: amount,
        direction: direction,
        counterparty: narration,
        reference: reference,
        narration: narration,
      ));
      previousBalance = balance;
    }

    return ChainValidated(transactions);
  }

  static String? _extractReference(String narration, List<RegExp> regexes) {
    for (final regex in regexes) {
      final match = regex.firstMatch(narration);
      if (match == null || match.groupCount < 1) continue;
      final captured = match.group(1)?.trim() ?? '';
      if (captured.isNotEmpty) return captured;
    }
    return null;
  }
}
